service/controller/resource/ipam/azure_machinepool_subnet_collector.go

```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	g := &errgroup.Group{}
+
+	g.Go(func() error {
+		// Collect subnets from AzureCluster CR.
+		azureClusterCRSubnets, err := c.collectSubnetsFromAzureClusterCR(ctx, azureCluster)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		mutex.Lock()
+		reservedSubnets = append(reservedSubnets, azureClusterCRSubnets...)
+		mutex.Unlock()
+
+		return nil
+	})
+
+	g.Go(func() error {
+		// Collect subnets from the actual Azure VNet via Azure API.
+		actualAzureVNetSubnets, err := c.collectSubnetsFromAzureVNet(ctx, azureCluster)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		mutex.Lock()
+		reservedSubnets = append(reservedSubnets, actualAzureVNetSubnets...)
+		mutex.Unlock()
+
+		return nil
+	})
+
+	err = g.Wait()
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// Check if subnets actually belong to the tenant cluster VNet and filter out the duplicates.
+	reservedSubnets = ipam.CanonicalizeSubnets(*tenantClusterVNetNetworkRange, reservedSubnets)
+
+	return reservedSubnets, nil
+}
+
+// collectSubnetsFromAzureClusterCR returns all subnets specified in AzureCluster CR.
+func (c *AzureMachinePoolSubnetCollector) collectSubnetsFromAzureClusterCR(_ context.Context, azureCluster *capzV1alpha3.AzureCluster) ([]net.IPNet, error) {
+	azureClusterCRSubnets := make([]net.IPNet, len(azureCluster.Spec.NetworkSpec.Subnets))
+
+	// Collect all the subnets from AzureCluster.Spec.NetworkSpec.Subnets field. If the Subnets
+	// field is not set, this function will simply return nil.
+	for _, subnet := range azureCluster.Spec.NetworkSpec.Subnets {
+		_, subnetIPNet, err := net.ParseCIDR(subnet.CidrBlock)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		azureClusterCRSubnets = append(azureClusterCRSubnets, *subnetIPNet)
+	}
+
+	return azureClusterCRSubnets, nil
+}
+
+// collectSubnetsFromAzureVNet returns all subnets that are deployed in Azure virtual network.
+func (c *AzureMachinePoolSubnetCollector) collectSubnetsFromAzureVNet(ctx context.Context, azureCluster *capzV1alpha3.AzureCluster) ([]net.IPNet, error) {
+	// Reads "giantswarm.io/organization" label from AzureCluster CR, and then uses organization
+	// name to get Azure credentials.
+	credentials, err := helpers.GetCredentialSecretFromMetadata(ctx, c.client, azureCluster.ObjectMeta)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	subnetsClient, err := c.azureClientFactory.GetSubnetsClient(credentials.Namespace, credentials.Name)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// cluster ID is tenant cluster resource group name
+	resourceGroupName := azureCluster.Name
+
+	// Not assuming VNet name here, keeping it flexible. In order to keep it correct and stable, we
+	// should have a webhook for enforcing a VNet name convention. This check can be removed once
+	// we have implemented a webhook.
```

Comment:
```suggestion
	// should have a webhook for enforcing a VNet name convention.
```
I don't think it hurts to have this here even if webhook validates it as well?

Reply:
It doesn't hurt, but I believe that if the webhook is validating the name and setting a default value, the name here will be always valid.

On the other side, if we do the defaulting in operator code, we need the check here as well :)